### PR TITLE
[docker-config-engine]: introduce docker sonic config engine

### PR DESCRIPTION
**Pull request by lguohan**

### CODE REVIEW

**Comment by lguohan** [comment]
one image size decreased by ~30%

```
-rw-r--r-- 1 lgh lgh 451227877 Feb  7 08:17 sonic-aboot-broadcom-1.swi
-rw-r--r-- 1 lgh lgh 615025670 Feb  7 02:58 sonic-aboot-broadcom.swi
```

File: dockers/docker-config-engine/Dockerfile.j2 [hunk 1]
```diff
@@ -0,0 +1,21 @@
+FROM docker-base
+
+RUN apt-get update
+
+# Dependencies for sonic-cfggen 
+RUN apt-get install -y python-lxml python-jinja2 python-netaddr python-ipaddr python-yaml 
+
+COPY \
+{% for deb in docker_config_engine_debs.split(' ') -%}
```
Comment (qiluo-msft) [inline dockers/docker-config-engine/Dockerfile.j2]: extra '-'?

File: dockers/docker-config-engine/Dockerfile.j2 [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+debs/{{ deb }}{{' '}}
+{%- endfor -%}
```
Comment (qiluo-msft) [inline dockers/docker-config-engine/Dockerfile.j2]: extra '-'s? {% endfor %}

File: dockers/docker-config-engine/Dockerfile.j2 (hunk 1 above)
Comment (lguohan) [inline dockers/docker-config-engine/Dockerfile.j2]: looks like it is the same as before

**Comment by lguohan** [comment]
rebase master

File: dockers/docker-config-engine/Dockerfile.j2 (hunk 1 above)
Comment (lguohan) [inline dockers/docker-config-engine/Dockerfile.j2]: You can also strip whitespace in templates by hand. If you add a minus sign (-) to the start or end of a block (e.g. a For tag), a comment, or a variable expression, the whitespaces before or after that block will be removed:

it is correct

File: dockers/docker-dhcp-relay/Dockerfile.j2 [hunk 3]
```diff
@@ -1,26 +1,11 @@
-FROM docker-base
+FROM docker-config-engine
 
 ## Make apt-get non-interactive
 ENV DEBIAN_FRONTEND=noninteractive
 
 ## Update APT package lists
 RUN apt-get update
 
-## Dependencies for sonic-cfggen
```
Comment (taoyl-ms) [inline dockers/docker-dhcp-relay/Dockerfile.j2]: I'll vote for keep those lines as removing them will cause confusion when later developers want to add dependency for this docker.
If we really want to remove those lines, maybe revert this docker from DOCKER_IMAGES back to SIMPLE_DOCKER_IMAGES in rule.mk?

File: dockers/docker-dhcp-relay/Dockerfile.j2 [hunk 4]
```diff
@@ -1,26 +1,11 @@
...
-RUN apt-get -y install python-lxml python-jinja2 python-netaddr python-ipaddr python-yaml
-
-
-COPY \
```
Comment (taoyl-ms) [inline dockers/docker-dhcp-relay/Dockerfile.j2]: I'll vote for keep those lines as removing them will cause confusion when later developers want to add dependency for this docker.
If we really want to remove those lines, maybe revert this docker from DOCKER_IMAGES back to SIMPLE_DOCKER_IMAGES in rule.mk?

File: dockers/docker-platform-monitor/Dockerfile.j2 [hunk 5]
```diff
@@ -1,24 +1,9 @@
...
+FROM docker-config-engine
 
 RUN apt-get update
 
 RUN apt-get install -y smartmontools sensord 
 
-# Dependencies for sonic-cfggen                                                                                                                                                           
-RUN apt-get install -y python-lxml python-jinja2 python-netaddr python-ipaddr python-yaml 
-
-COPY \
```
Comment (taoyl-ms) [inline dockers/docker-platform-monitor/Dockerfile.j2]: Same with dhcp-relay. Revert to SIMPLE_DOCKER_IMAGES?

File: dockers/docker-dhcp-relay/Dockerfile.j2 (hunk 4 above)
Comment (lguohan) [inline dockers/docker-dhcp-relay/Dockerfile.j2]: will do.

File: dockers/docker-platform-monitor/Dockerfile.j2 (hunk 5 above)
Comment (lguohan) [inline dockers/docker-platform-monitor/Dockerfile.j2]: will do.

File: dockers/docker-dhcp-relay/Dockerfile.j2 (hunk 4 above)
Comment (lguohan) [inline dockers/docker-dhcp-relay/Dockerfile.j2]: updated

**Comment by taoyl-ms** [comment]
:shipit:

**Comment by taoyl-ms** [comment]
:clock1:

File: dockers/docker-dhcp-relay/Dockerfile.j2 (hunk 4 above)
Comment (taoyl-ms) [inline dockers/docker-dhcp-relay/Dockerfile.j2]: Should rename from Dockerfile.j2 to Dockerfile as well? SIMPLE_DOCKER don't use dockerfile templates

Comment (lguohan) [inline dockers/docker-dhcp-relay/Dockerfile.j2]: fixed

**Pull request merged**